app/controllers/schools/add_participants_controller.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module Schools
+  class AddParticipantsController < ::Schools::BaseController
+    FORM_SESSION_KEY = :add_participant_form
+    FORM_PARAM_KEY = :schools_add_participant_form
+
+    skip_after_action :verify_authorized
+    before_action :ensure_form_present, except: :start
+    before_action :set_school_cohort
+
+    helper_method :add_participant_form
+
+    def start
+      session.delete(FORM_SESSION_KEY)
+      session[FORM_SESSION_KEY] = {
+        school_cohort_id: @school_cohort.id,
+      }
+      redirect_to action: :show, step: :type
+    end
+
+    def show
+      render current_step
+    end
+
+    def update
+      if add_participant_form.valid?(current_step)
+        add_participant_form.record_completed_step current_step
+        store_form_in_session
+        redirect_to action: :show, step: step_param(add_participant_form.next_step(current_step))
+      else
+        render current_step
+      end
+    end
+
+    def complete
+      @participant = add_participant_form.save!
+      @type = add_participant_form.type
+
+      session.delete(FORM_SESSION_KEY)
+    end
+
+  private
+
+    def add_participant_form
+      return @add_participant_form if defined?(@add_participant_form)
+
+      @add_participant_form = AddParticipantForm.new(session[FORM_SESSION_KEY])
+      @add_participant_form.assign_attributes(add_participant_form_params) if params[FORM_PARAM_KEY]
+
+      @add_participant_form
+    end
+
+    def store_form_in_session
+      session[FORM_SESSION_KEY] = add_participant_form.attributes
+    end
+
+    def current_step
+      params[:step].tr("-", "_").to_sym
+    end
+
+    def step_param(step)
+      step.to_s.tr("_", "-")
+    end
+
+    def back_link_path
+      previous_step = add_participant_form.previous_step(current_step)
+      return schools_cohort_participants_path unless previous_step
+
+      { action: :show, step: step_param(previous_step) }
+    end
+
+    def add_participant_form_params
+      params.require(FORM_PARAM_KEY).permit(:type, :full_name, :email, :mentor_id)
+    end
+
+    def email_used_in_the_same_school?
+      User.find_by(email: add_participant_form.email).school == add_participant_form.school_cohort.school
+    end
```

Comment:
Yeah, I was on edge with that one. Ideally, I'd convert our step views into view-components and then that's where this code should be then. For now, neither form nor controller seems to be a great place for it. Alternatively, we could split the `email_taken` view into two separate view, and then have this check within forms `next_step` logic.

Actually there are many alternatives - previously @callumacrae has wrote this check as an additional validation on the form and some extra code in the controller to pick up that specific error and display the email_taken page instead. I've changed this into having `email_taken` as a step instead as, to me, at the point that page is displayed the details step has been completed, resulting in the premature termination of the flow. 

Yet another option would be to keep it as a validation, but make it strict and rescue the exception in the controller. This would however require a special treatmant of `previous_step` as it will preven `details` step from completion which does not feel right.